File: python/app/application/user_query_handler.py
from abc import ABC, abstractmethod

from app.application.user_dto import UserDto
from app.application.user_query_criteria import UserQueryCriteria
from app.application.user_query_sort import UserQuerySort


class UserQueryHandler(ABC):
    """用户Repository接口"""

    @abstractmethod
    def query_by_id(self, id: str) -> UserDto | None:
        pass

    @abstractmethod
    def query_by_id_req(self, id: str) -> UserDto:
        pass

    @abstractmethod
    def query_count(self, criteria: UserQueryCriteria) -> int:
        pass

    @abstractmethod
    def query(self, criteria: UserQueryCriteria, *sorts: UserQuerySort) -> list[UserDto]:
        pass

    @abstractmethod
    def query_by_page(self, page_num: int, page_size: int, criteria: UserQueryCriteria, *sorts: UserQuerySort) -> list[UserDto]:
        pass

File: dotnet/src/application/UserQueryHandler.cs
namespace DddExample.Application
{
    /// <summary>
    /// 用户 Query Handler 接口
    /// </summary>
    public interface UserQueryHandler
    {
        /// <summary>
        /// 根据 id 查询，找不到返回 null
        /// </summary>
        UserDto? QueryById(string id);

        /// <summary>
        /// 根据 id 查询，找不到抛出异常
        /// </summary>
        UserDto QueryByIdReq(string id);

        /// <summary>
        /// 查询记录数
        /// </summary>
        int QueryCount(UserQueryCriteria? criteria);

        /// <summary>
        /// 查询
        /// </summary>
        IList<UserDto> Query(UserQueryCriteria? criteria, params UserQuerySort[] sorts);

        /// <summary>
        /// 分页查询
        /// </summary>
        IList<UserDto> QueryByPage(int pageNum, int pageSize, UserQueryCriteria? criteria, params UserQuerySort[] sorts);
    }
}
